internal/operator/clientservice/clientservice.go

```diff
@@ -0,0 +1,174 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package clientservice
+
+import (
+	"fmt"
+
+	druidv1alpha1 "github.com/gardener/etcd-druid/api/v1alpha1"
+	"github.com/gardener/etcd-druid/internal/common"
+	druiderr "github.com/gardener/etcd-druid/internal/errors"
+	"github.com/gardener/etcd-druid/internal/operator/component"
+	"github.com/gardener/etcd-druid/internal/utils"
+	"github.com/gardener/gardener/pkg/controllerutils"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// ErrGetClientService indicates an error in getting the client service resource.
+	ErrGetClientService druidv1alpha1.ErrorCode = "ERR_GET_CLIENT_SERVICE"
+	// ErrSyncClientService indicates an error in syncing the client service resource.
+	ErrSyncClientService druidv1alpha1.ErrorCode = "ERR_SYNC_CLIENT_SERVICE"
+	// ErrDeleteClientService indicates an error in deleting the client service resource.
+	ErrDeleteClientService druidv1alpha1.ErrorCode = "ERR_DELETE_CLIENT_SERVICE"
+)
+
+type _resource struct {
+	client client.Client
+}
+
+// New returns a new client service operator.
+func New(client client.Client) component.Operator {
+	return &_resource{
+		client: client,
+	}
+}
+
+// GetExistingResourceNames returns the name of the existing client service for the given Etcd.
+func (r _resource) GetExistingResourceNames(ctx component.OperatorContext, etcd *druidv1alpha1.Etcd) ([]string, error) {
+	resourceNames := make([]string, 0, 1)
+	svc := &corev1.Service{}
+	svcObjectKey := getObjectKey(etcd)
+	if err := r.client.Get(ctx, svcObjectKey, svc); err != nil {
```

Comment:
This could be metadata-only get instead of the full object.

Reply:
change has been made. Now we use `PartialObjectMetadata` to fetch only the metadata of an object instead of fetching the entire object in `GetExistingResourceNames` method implementations.